src/libopensc/card-cac1.c

```diff
@@ -0,0 +1,558 @@
+/*
+ * card-cac1.c: Support for legacy CAC-1
+ * card-default.c: Support for cards with no driver
+ *
+ * Copyright (C) 2001, 2002  Juha Yrjölä <juha.yrjola@iki.fi>
+ * Copyright (C) 2005,2006,2007,2008,2009,2010 Douglas E. Engert <deengert@anl.gov>
+ * Copyright (C) 2006, Identity Alliance, Thomas Harning <thomas.harning@identityalliance.com>
+ * Copyright (C) 2007, EMC, Russell Larner <rlarner@rsa.com>
+ * Copyright (C) 2016 - 2018, Red Hat, Inc.
+ *
+ * CAC driver author: Robert Relyea <rrelyea@redhat.com>
+ * Further work: Jakub Jelen <jjelen@redhat.com>
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+ */
+
+#if HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include <ctype.h>
+#include <fcntl.h>
+#include <limits.h>
+#include <stdlib.h>
+#include <string.h>
+
+#ifdef _WIN32
+#include <io.h>
+#else
+#include <unistd.h>
+#endif
+
+#ifdef ENABLE_OPENSSL
+#include <openssl/sha.h>
+#endif /* ENABLE_OPENSSL */
+
+#include "internal.h"
+#include "simpletlv.h"
+#include "cardctl.h"
+#ifdef ENABLE_ZLIB
+#include "compression.h"
+#endif
+#include "iso7816.h"
+#include "card-cac-common.h"
+
+/*
+ *  CAC hardware and APDU constants
+ */
+#define CAC_INS_GET_CERTIFICATE       0x36  /* CAC1 command to read a certificate */
+
+/*
+ * OLD cac read certificate, only use with CAC-1 card.
+ */
+static int cac_cac1_get_certificate(sc_card_t *card, u8 **out_buf, size_t *out_len)
+{
+	u8 buf[CAC_MAX_SIZE];
+	u8 *out_ptr;
+	size_t size = 0;
+	size_t left = 0;
+	size_t len, next_len;
+	sc_apdu_t apdu;
+	int r = SC_SUCCESS;
+	SC_FUNC_CALLED(card->ctx, SC_LOG_DEBUG_VERBOSE);
+	/* get the size */
+	size = left = *out_buf ? *out_len : sizeof(buf);
+	out_ptr = *out_buf ? *out_buf : buf;
+	sc_format_apdu(card, &apdu, SC_APDU_CASE_2_SHORT, CAC_INS_GET_CERTIFICATE, 0, 0 );
+	next_len = MIN(left, 100);
+	for (; left > 0; left -= len, out_ptr += len) {
```

Comment:
I think this needs error checking so that the response isn't written beyond `buf`/`*out_buf`

Reply:
I do not see how this could overflow. Is the apdu code able to write behind the `LE` boundary? The `apdu.le` is set to `len`, which is set from `next_len`, which properly checks against the maximum bound defined in the `left` variable.

Reply:
I think you're right. But at least there's the possibility that this loops forever, if the card returns 6300, which sets next_len to 0

Reply:
You are right. Fixed in the next commit.